Manual feature configuration

planarvoid opened the pull request:
This PR adds functionality to manually override any local or remote feature flag. I know this PR is big but as this feature will only ever be used by us or testers, I thought it would be OK (but I can still split the PR into more if necessary). 

I've added a `ManualFeatureConfig` which saves the locally used feature state into apps preferences. This functionality is behind a flag which is available in wasabi and jalapeno. The main reason is that we want to be able to test a developed feature on a PR build. I tried to minimize the amount of changes by reusing existing styles/dimens (cause the UI of this screen doesn't really matter).

The new screen can be found in Me/App settings/Test feature configuration if you're on Jalapeno or Wasabi builds. The screen has 2 sections. One for the remote features (the list is generated from the feature config files annotated with the `Feature` annotation) and the other for features in development (the list is generated from the feature config files annotated with the new `FeatureInDevelopment` annotation). You might notice that the features in development start in a `?` state. The reason is that we only show the manually overridden value there and there is no way to find out the current build config value without affecting the dev comfort while creating the flags.

I've added a "Restart" button which appears after any change. The reason is that the changes to flags sometimes need the app to be restarted to appear. The button should nudge the devs/testers to do so. 

![Screenshot_1600173924](https://user-images.githubusercontent.com/1079756/93212818-649c2880-f763-11ea-93e7-c3110dd14902.png)


To test:
- Go to Me/App settings/Test feature configuration
- Notice the new screen
- Notice that the remote configs are set based on the remote value
- Notice that the features in development are all `?`
- Try enabling/disabling features
- Check that this state is stored and survives app restart
- Check that the state has impact on the features
- Check that the restart button appears on any change

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid committed:
Introduce manual feature config for testing purposes

planarvoid committed:
Add FeatureInDevelopment.kt annotation and processing

planarvoid committed:
Update manual feature config screen to use generated features in development

planarvoid committed:
Fix lint issue

planarvoid committed:
Rename setup package to manual

planarvoid committed:
Add tests for app config and manual config

planarvoid committed:
Add test for the ManualFeatureConfigViewModel.kt

planarvoid committed:
Fix lint issues

malinajirka commented on WordPress/build.gradle:
```diff
@@ -102,6 +103,7 @@ android {
             buildConfigField "boolean", "MODAL_LAYOUT_PICKER", "false"
             buildConfigField "boolean", "WP_STORIES_AVAILABLE", "false"
             buildConfigField "boolean", "ANY_FILE_UPLOAD", "false"
+            buildConfigField "boolean", "ENABLE_FEATURE_CONFIGURATION", "false"
```
Do we really want to keep this feature enabled in "zalpha"?

planarvoid replied:
Since the description of zalpha is: 
```
zalpha { // alpha version - enable experimental features
```
I thought it fits in there 🤷 , I'm fine with disabling it. Who gets zalpha?

malinajirka replied:
Alpha is used for closed testing track distributed through Google Play. However, since the purpose of this screen is to allow devs+testers test features which are not prepared for a release yet I believe we don't need to have this screen in internal testing phase. We might want to consider disabling it there for now and enable it if we find the need. Wdyt?

planarvoid replied:
makes sense. I'll disable it here and leave it on only for wasabi and jalapeno

malinajirka commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -875,6 +875,13 @@
         <item>@string/app_theme_entry_value_dark</item>
     </string-array>
 
+    <!-- Feature configuration -->
```
We might want to consider marking these strings as `translatable=false`. I don't think we need translations for them. Wdyt?

malinajirka reviewed (changes requested):
Thanks @planarvoid! The changes look good overall. However, I tried to set `ENABLE_FEATURE_CONFIGURATION` to false and run the app (to simulate a release build). And the app crashes with the following exception:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android.beta, PID: 18119
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.wordpress.android.beta/org.wordpress.android.ui.prefs.AppSettingsActivity}: android.view.InflateException: Binary XML file line #9: Binary XML file line #9: Error inflating class fragment
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: android.view.InflateException: Binary XML file line #9: Binary XML file line #9: Error inflating class fragment
     Caused by: android.view.InflateException: Binary XML file line #9: Error inflating class fragment
     Caused by: java.lang.ClassCastException: android.preference.Preference cannot be cast to android.preference.PreferenceCategory
        at org.wordpress.android.ui.prefs.AppSettingsFragment.removeFeatureConfigCategory(AppSettingsFragment.java:203)
        at org.wordpress.android.ui.prefs.AppSettingsFragment.onCreate(AppSettingsFragment.java:189)
        at android.app.Fragment.performCreate(Fragment.java:2489)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1237)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1462)
        at android.app.FragmentManagerImpl.addFragment(FragmentManager.java:1704)
        at android.app.FragmentManagerImpl.onCreateView(FragmentManager.java:3537)
        at android.app.FragmentController.onCreateView(FragmentController.java:98)
        at android.app.Activity.onCreateView(Activity.java:6219)
        at androidx.fragment.app.FragmentActivity.onCreateView(FragmentActivity.java:337)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:780)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:730)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:863)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:824)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:515)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
        at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:555)
        at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:161)
        at org.wordpress.android.ui.prefs.AppSettingsActivity.onCreate(AppSettingsActivity.java:19)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

planarvoid committed:
Fix AppSettingsFragment to remove manual feature config when flag is disabled

planarvoid committed:
Make strings not translatable

planarvoid commented:
Thanks for the review @malinajirka! I've fixed the crash in e254768 and made the strings not translatable in 632dd7e. I've also renamed the strings to something specific to this usecase so someone doesn't accidentally uses them in a translated part of the app.

malinajirka reviewed (approved):
LGTM, thanks! Feel free to merge it no matter what you decided to do with the alpha flavor.

planarvoid committed:
Disable feature configuration in zalpha flavor
